modules/deser-lib/src/cbor.ts

```diff
@@ -0,0 +1,217 @@
+ * Type check for sortable array element.
+ * @param value - Value to type check.
+ * @returns True if value is a sortable array element.
+ */
+function isSortable(value: unknown): value is Sortable {
+  return value instanceof Object && 'weight' in value;
+}
+
+/**
+ * Convert number to base 256 and return as a big-endian Buffer.
+ * @param value - Value to convert.
+ * @returs Buffer representation of the number.
+ */
+function numberToBufferBE(value: number): Buffer {
+  // Normalize value so that negative numbers aren't compared higher
+  // than positive numbers when accounting for two's complement.
+  value += Math.pow(2, 52);
+  const byteCount = Math.floor((value.toString(2).length + 7) / 8);
+  const buffer = Buffer.alloc(9);
+  const offset = 8 - byteCount;
+  let i = 0;
+  while (value) {
+    buffer[offset + i++] = value % 256;
+    value = Math.floor(value / 256);
+  }
+  return buffer.reverse();
+}
+
+/**
+ * Compare two array elements for sorting.
+ * @param a - left element to compare to right element.
+ * @param b - right element to compare to left element.
+ * @returns Negative if a < b, positive if b > a, 0 if equal.
+ */
+function elementCompare(a: unknown, b: unknown): number {
+  if (!isSortable(a) || !isSortable(b)) {
+    throw new Error('Array elements must be sortable');
+  }
+  if (a.weight === b.weight) {
+    if ('value' in a && 'value' in b) {
+      const aVal = transform(a.value);
+      const bVal = transform(b.value);
+      if (
+        (!Buffer.isBuffer(aVal) && typeof aVal !== 'string' && typeof aVal !== 'number') ||
+        (!Buffer.isBuffer(bVal) && typeof bVal !== 'string' && typeof bVal !== 'number')
+      ) {
+        throw new Error('Array element value cannot be compared');
+      }
+      let aBuf, bBuf;
+      if (typeof aVal === 'number') {
+        aBuf = numberToBufferBE(aVal);
+      } else {
+        aBuf = Buffer.from(aVal);
+      }
+      if (typeof bVal === 'number') {
+        bBuf = numberToBufferBE(bVal);
+      } else {
+        bBuf = Buffer.from(bVal);
+      }
+      return bufferCompare(aBuf, bBuf);
+    }
+    throw new Error('Array elements must be sortable');
+  }
+  return a.weight - b.weight;
+}
+
+/**
+ * Transform value into its canonical, serializable form.
+ * @param value - Value to transform.
+ * @returns Canonical, serializable form of value.
+ */
+export function transform<T>(value: T): T | Buffer {
+  if (value === null || value === undefined) {
+    return value;
+  }
+  if (typeof value === 'string') {
+    // Transform hex strings to buffers.
+    if (value.startsWith('0x')) {
+      if (!value.match(/^0x([0-9a-fA-F]{2})*$/)) {
+        throw new Error('0x prefixed string contains non-hex characters.');
+      }
+      return Buffer.from(value.slice(2), 'hex');
+    } else if (value.startsWith('\\0x')) {
+      return value.slice(1) as unknown as T;
+    }
+  } else if (value instanceof Array) {
+    // Enforce array elements are same type.
+    getType(value);
```

Comment:
I think you missed something here, otherwise there is no use for the `getType` function

Reply:
getType enforces that array elements are of the same type. If there are elements of different types, then getType will throw, which is what we want since we cannot canonically sort different types of objects.